wrapper/src/main/java/software/amazon/jdbc/hostlistprovider/monitoring/GlobalDbClusterTopologyMonitorImpl.java

```diff
@@ -0,0 +1,109 @@
+ * limitations under the License.
+ */
+
+package software.amazon.jdbc.hostlistprovider.monitoring;
+
+import java.sql.Connection;
+import java.sql.PreparedStatement;
+import java.sql.ResultSet;
+import java.sql.SQLException;
+import java.sql.Timestamp;
+import java.time.Instant;
+import java.util.List;
+import java.util.Map;
+import java.util.Properties;
+import java.util.logging.Logger;
+import software.amazon.jdbc.HostListProviderService;
+import software.amazon.jdbc.HostSpec;
+import software.amazon.jdbc.PluginService;
+import software.amazon.jdbc.util.CacheMap;
+import software.amazon.jdbc.util.StringUtils;
+
+public class GlobalDbClusterTopologyMonitorImpl extends ClusterTopologyMonitorImpl {
+
+  private static final Logger LOGGER = Logger.getLogger(GlobalDbClusterTopologyMonitorImpl.class.getName());
+
+  protected final Map<String, HostSpec> globalClusterInstanceTemplateByAwsRegion;
+  protected final String regionByNodeIdQuery;
+
+  public GlobalDbClusterTopologyMonitorImpl(String clusterId,
+      CacheMap<String, List<HostSpec>> topologyMap,
+      HostSpec initialHostSpec, Properties properties,
+      PluginService pluginService,
+      HostListProviderService hostListProviderService,
+      HostSpec clusterInstanceTemplate, long refreshRateNano,
+      long highRefreshRateNano, long topologyCacheExpirationNano, String globalTopologyQuery,
+      String writerTopologyQuery, String nodeIdQuery,
+      Map<String, HostSpec> globalClusterInstanceTemplateByAwsRegion,
+      String regionByNodeIdQuery) {
+
+    super(clusterId, topologyMap, initialHostSpec, properties, pluginService, hostListProviderService,
+        clusterInstanceTemplate, refreshRateNano, highRefreshRateNano, topologyCacheExpirationNano, globalTopologyQuery,
+        writerTopologyQuery, nodeIdQuery);
+    this.globalClusterInstanceTemplateByAwsRegion = globalClusterInstanceTemplateByAwsRegion;
+    this.regionByNodeIdQuery = regionByNodeIdQuery;
+  }
+
+  @Override
+  protected HostSpec getClusterInstanceTemplate(String nodeId, Connection connection) {
+    try {
+      try (final PreparedStatement stmt = connection.prepareStatement(this.regionByNodeIdQuery)) {
+        stmt.setString(1, nodeId);
+        try (final ResultSet resultSet = stmt.executeQuery()) {
+          if (resultSet.next()) {
+            String awsRegion = resultSet.getString(1);
+            if (!StringUtils.isNullOrEmpty(awsRegion)) {
+              final HostSpec clusterInstanceTemplateForRegion
+                  = this.globalClusterInstanceTemplateByAwsRegion.get(awsRegion);
+              if (clusterInstanceTemplateForRegion == null) {
+                throw new SQLException("Can't find cluster template for region " + awsRegion);
+              }
+              return clusterInstanceTemplateForRegion;
+            }
+          }
+        }
+      }
+    } catch (SQLException ex) {
+      throw new RuntimeException(ex);
+    }
+    return this.clusterInstanceTemplate;
+  }
```

Comment:
To get lower the level of nesting:
```suggestion
protected HostSpec getClusterInstanceTemplate(String nodeId, Connection connection) {
    try (final PreparedStatement stmt = connection.prepareStatement(this.regionByNodeIdQuery)) {
      stmt.setString(1, nodeId);
      try (final ResultSet resultSet = stmt.executeQuery()) {
        if (!resultSet.next()) {
          return this.clusterInstanceTemplate;
        }
        
        String awsRegion = resultSet.getString(1);
        if (StringUtils.isNullOrEmpty(awsRegion)) {
          return this.clusterInstanceTemplate;
        }

        final HostSpec clusterInstanceTemplateForRegion = this.globalClusterInstanceTemplateByAwsRegion.get(awsRegion);
        if (clusterInstanceTemplateForRegion == null) {
          throw new SQLException("Can't find cluster template for region " + awsRegion);
        }
        return clusterInstanceTemplateForRegion;
      }
    } catch (SQLException ex) {
      throw new RuntimeException(ex);
    }
  }
```